File: scrapers/natepann.py
import re

import httpx

from scrapers.base import BaseScraper, Post


class NatepannScraper(BaseScraper):
    community = "natepann"
    community_name = "네이트판"
    base_url = "https://pann.nate.com"
    data_required = True

    async def fetch_best_posts(self, client: httpx.AsyncClient) -> list[Post]:
        soup = await self._fetch_html(client, f"{self.base_url}/talk/ranking")
        posts = []

        for li in soup.select("ul.post_wrap > li"):
            a_tag = li.select_one("h2 a[href]")
            if not a_tag:
                continue
            title = a_tag.get_text(strip=True)
            href = a_tag.get("href", "")
            if not title or not href:
                continue

            votes = 0
            rcm = li.select_one("span.rcm")
            if rcm:
                m = re.search(r"\d+", rcm.get_text())
                if m:
                    votes = int(m.group())

            comments = 0
            reple = li.select_one("span.reple-num")
            if reple:
                m = re.search(r"\d+", reple.get_text())
                if m:
                    comments = int(m.group())

            url = f"{self.base_url}{href}" if not href.startswith("http") else href
            posts.append(self._make_post(title, url, votes, comments=comments))

        # 기존 방식 폴백
        if not posts:
            for a_tag in soup.find_all("a", href=re.compile(r"^/talk/\d+")):
                if a_tag.parent and a_tag.parent.name == "h2":
                    title = a_tag.get_text(strip=True)
                    href = a_tag["href"]
                    if not title:
                        continue
                    url = f"{self.base_url}{href}"
                    posts.append(self._make_post(title, url))

        return posts

File: internal/scraper/natepann.go
package scraper

import (
	"fmt"
	"net/http"
	"regexp"
	"strconv"
	"strings"

	"github.com/PuerkitoBio/goquery"
)

var natepannDigitRe = regexp.MustCompile(`\d+`)
var natepannFallbackRe = regexp.MustCompile(`^/talk/\d+`)

// NatepannScraper scrapes 네이트판 ranking.
type NatepannScraper struct {
	baseScraper
}

func NewNatepannScraper() *NatepannScraper {
	return &NatepannScraper{
		baseScraper: baseScraper{
			community:     "natepann",
			communityName: "네이트판",
			baseURL:       "https://pann.nate.com",
			encoding:      "utf-8",
			dataRequired:  true,
		},
	}
}

func (s *NatepannScraper) Name() string { return s.communityName }

func (s *NatepannScraper) FetchBestPosts(client *http.Client) ([]Post, error) {
	doc, err := fetchDocument(client, fmt.Sprintf("%s/talk/ranking", s.baseURL), s.encoding, s.baseURL)
	if err != nil {
		return nil, err
	}

	var posts []Post

	doc.Find("ul.post_wrap > li").Each(func(_ int, li *goquery.Selection) {
		aTag := li.Find("h2 a[href]").First()
		if aTag.Length() == 0 {
			return
		}
		title := strings.TrimSpace(aTag.Text())
		href, exists := aTag.Attr("href")
		if !exists || title == "" || href == "" {
			return
		}

		votes := 0
		rcm := li.Find("span.rcm").First()
		if rcm.Length() > 0 {
			m := natepannDigitRe.FindString(rcm.Text())
			if m != "" {
				if v, err := strconv.Atoi(m); err == nil {
					votes = v
				}
			}
		}

		comments := 0
		reple := li.Find("span.reple-num").First()
		if reple.Length() > 0 {
			m := natepannDigitRe.FindString(reple.Text())
			if m != "" {
				if v, err := strconv.Atoi(m); err == nil {
					comments = v
				}
			}
		}

		url := href
		if !strings.HasPrefix(href, "http") {
			url = s.baseURL + href
		}
		posts = append(posts, s.makePost(title, url, votes, 0, comments))
	})

	// Fallback: broader selector
	if len(posts) == 0 {
		doc.Find("a[href]").Each(func(_ int, aTag *goquery.Selection) {
			href, _ := aTag.Attr("href")
			if !natepannFallbackRe.MatchString(href) {
				return
			}
			// Check parent is h2
			parent := aTag.Parent()
			if parent.Length() == 0 || goquery.NodeName(parent) != "h2" {
				return
			}
			title := strings.TrimSpace(aTag.Text())
			if title == "" {
				return
			}
			url := s.baseURL + href
			posts = append(posts, s.makePost(title, url, 0, 0, 0))
		})
	}

	return s.filterPosts(posts), nil
}
